bugsnag-plugin-android-anr/src/main/jni/anr_handler.c

```diff
@@ -5,118 +5,115 @@
+
+static pthread_t watchdog_thread;
+static bool should_report_anr = false;
+static struct sigaction original_sigquit_handler;
 
 static JavaVM *bsg_jvm = NULL;
 static jmethodID mthd_notify_anr_detected = NULL;
 static jobject obj_plugin = NULL;
 
-/* The previous SIGQUIT handler */
-struct sigaction bsg_sigquit_sigaction_previous;
-
-void bsg_notify_anr_detected(JNIEnv *env) {
-  (*env)->CallVoidMethod(env, obj_plugin, mthd_notify_anr_detected);
-}
-
-bool bsg_configure_anr_jni(JNIEnv *env) {
+static bool configure_anr_jni(JNIEnv *env) {
   // get a global reference to the AnrPlugin class
   // https://developer.android.com/training/articles/perf-jni#faq:-why-didnt-findclass-find-my-class
   int result = (*env)->GetJavaVM(env, &bsg_jvm);
   if (result != 0) {
-    // Failed to fetch VM, cannot continue
+    BUGSNAG_LOG("Failed to fetch Java VM. ANR handler not installed.");
     return false;
   }
 
   jclass clz = (*env)->FindClass(env, "com/bugsnag/android/AnrPlugin");
   mthd_notify_anr_detected =
-      (*env)->GetMethodID(env, clz, "notifyAnrDetected", "()V");
+          (*env)->GetMethodID(env, clz, "notifyAnrDetected", "()V");
   return true;
 }
 
-void bsg_handle_sigquit(int signum, siginfo_t *info, void *user_context) {
+static void notify_anr_detected() {
   if (enabled) {
-    // invoke a JNI call from the SIGQUIT handler
     JNIEnv *env;
-    int result = (*bsg_jvm)->GetEnv(bsg_jvm, (void **)&env, JNI_VERSION_1_4);
+    int result = (*bsg_jvm)->GetEnv(bsg_jvm, (void **) &env, JNI_VERSION_1_4);
 
     if (result == JNI_OK) { // already attached
-      bsg_notify_anr_detected(env);
+      (*env)->CallVoidMethod(env, obj_plugin, mthd_notify_anr_detected);
     } else if (result == JNI_EDETACHED) { // attach before calling JNI
       if ((*bsg_jvm)->AttachCurrentThread(bsg_jvm, &env, NULL) == 0) {
-        bsg_notify_anr_detected(env);
+        (*env)->CallVoidMethod(env, obj_plugin, mthd_notify_anr_detected);
         (*bsg_jvm)->DetachCurrentThread(
-            bsg_jvm); // detach to restore initial condition
+                bsg_jvm); // detach to restore initial condition
       }
     } // All other results are error codes
   }
+}
 
-  // Invoke previous handler, if a custom handler has been set
-  // This can be conditionally switched off on certain platforms (e.g. Unity)
-  // where this behaviour is not desirable due to Unity's implementation failing
-  // with a SIGSEGV
-  if (invokePrevHandler) {
-    struct sigaction previous = bsg_sigquit_sigaction_previous;
-    if (previous.sa_flags & SA_SIGINFO) {
-      previous.sa_sigaction(SIGQUIT, info, user_context);
-    } else if (previous.sa_handler == SIG_DFL) {
-      // Do nothing, the default action is nothing
-    } else if (previous.sa_handler != SIG_IGN) {
-      void (*previous_handler)(int) = previous.sa_handler;
-      previous_handler(signum);
-    }
+static void* sigquit_thread_main(void* _) {
+  while(!should_report_anr) {
+    usleep(100000);
```

Comment:
It would be good to extract this as a constant or add some documentation about why it's been selected - the context is missing for future readers of a 10s gap being required to avoid process termination.